File: Server/src/main/java/com/github/dr/rwserver/net/game/TimeoutDetection.java
package com.github.dr.rwserver.net.game;

import com.github.dr.rwserver.net.core.AbstractNetConnect;
import com.github.dr.rwserver.util.Time;
import com.github.dr.rwserver.util.threads.ThreadFactoryName;

import java.util.concurrent.*;

/**
 * @author Dr
 */
class TimeoutDetection {
    private final ThreadFactory namedFactory = ThreadFactoryName.nameThreadFactory("TimeoutDetection-");
    private final ScheduledExecutorService SERVICE = new ScheduledThreadPoolExecutor(1,namedFactory);
    private final ScheduledFuture scheduledFuture;

    protected TimeoutDetection(int s,final StartNet startNet) {
        scheduledFuture = SERVICE.scheduleAtFixedRate(new CheckTime(startNet),0,s, TimeUnit.SECONDS);
    }

    protected static boolean checkTimeoutDetection(final StartNet startNet, final AbstractNetConnect abstractNetConnect) {
        if (abstractNetConnect == null) {
            return true;
        }
        if (abstractNetConnect.getIsPasswd()) {
            /* 60s无反应判定close */
            return Time.concurrentMillis() > (abstractNetConnect.getLastReceivedTime() + 60 * 1000L);
        } else {
            return Time.concurrentMillis() > (abstractNetConnect.getLastReceivedTime() + 180 * 1000L);
        }
    }

    private static class CheckTime implements Runnable {
        private final StartNet startNet;

        private CheckTime(StartNet startNet) {
            this.startNet = startNet;
        }
        @Override
        public void run() {
            startNet.OVER_MAP.each((k,v) -> {
                if (checkTimeoutDetection(startNet,v)) {
                    v.disconnect();
                    startNet.OVER_MAP.remove(k);
                }
            });
        }
    }

}

File: Server/src/main/java/com/github/dr/rwserver/net/game/TimeoutDetection.kt
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package com.github.dr.rwserver.net.game

import com.github.dr.rwserver.net.core.TypeConnect
import com.github.dr.rwserver.net.core.server.AbstractNetConnect
import com.github.dr.rwserver.util.Time.concurrentMillis
import com.github.dr.rwserver.util.threads.ThreadFactoryName
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit

/**
 * @author Dr
 */
internal class TimeoutDetection(s: Int, startNet: StartNet) {
    private val namedFactory = ThreadFactoryName.nameThreadFactory("TimeoutDetection-")
    private val Service: ScheduledExecutorService = ScheduledThreadPoolExecutor(1, namedFactory)
    private val scheduledFuture: ScheduledFuture<*> = Service.scheduleAtFixedRate(CheckTime(startNet), 0, s.toLong(), TimeUnit.SECONDS)

    private class CheckTime(private val startNet: StartNet) : Runnable {
        override fun run() {
            startNet.OVER_MAP.each { k: String, v: TypeConnect ->
                val con = v.abstractNetConnect
                if (checkTimeoutDetection(con)) {
                    con.disconnect()
                    startNet.OVER_MAP.remove(k)
                }
            }
        }
    }

    companion object {
        internal fun checkTimeoutDetection(abstractNetConnect: AbstractNetConnect?): Boolean {
            if (abstractNetConnect == null) {
                return true
            }

            return if (abstractNetConnect.inputPassword) {
                /* 60s无反应判定close */
                concurrentMillis() > abstractNetConnect.lastReceivedTime + 60 * 1000L
            } else {
                concurrentMillis() > abstractNetConnect.lastReceivedTime + 180 * 1000L
            }

        }
    }
}